src/java/org/apache/cassandra/config/Config.java

```diff
@@ -359,9 +352,14 @@ public MemtableOptions()
 
     public String[] data_file_directories = new String[0];
 
-    public List<ParameterizedClass> sstable_formats = ImmutableList.of(new ParameterizedClass(BigFormat.class.getName(),// "org.apache.cassandra.io.sstable.format.big.BigFormat",
-                                                                                              ImmutableMap.of(SSTABLE_FORMAT_ID, "0",
-                                                                                                              SSTABLE_FORMAT_NAME, "big")));
+    public static class SSTableFormats
+    {
+        public String selected_format;
+        public String selected_version;
```

Comment:
my thinking is, if we do want to later (not this patch) allow overriding via JMX or Settings table, how can you do thing safely?

Lets say we have the following cases

1) keep same format, just change version (use case; major/minor upgrade wants to delay using latest until "stable")

```
sstable_formats.selected_version = next // this is "safe"
```

2) switch from big to bti or some 3rd party format

```
sstable_formats.selected_version = bti_version // unsafe, breaks BIG
sstable_formats.selected_format = bti
```

or

```
sstable_formats.selected_format = bti // unsafe as "selected_version" still points to BIG version
sstable_formats.selected_version = bti_version
```

Reply:
Where would we express the expected target format and version then? If we ever go for changing that at runtime, this should always be changed together.

Reply:
> this should always be changed together

agree, but we would need locking or copy-on-write configs to support that, which we don't have in this patch (and not arguing we add).  This means a config change happens each time a field is modified, which could put us in a inconsistence state waiting for the operator to update the other fields.

So, by having the version in `options` this at least scopes things to where they make sense... if you are using `big` and want to test out `bti v42` you can update the configs for `bti.version=42`, this would have no runtime impact as its untouched.  But let's say you are using `bti v41` and want to switch to `v42`?  you then do `bit.version=42` and the next sstable uses `v42`.  By having the version in the options it makes the concurrency easier to reason about

Reply:
agreed; though I was also thinking about specifying the write format like:

```
write_format: big
```
or
```
write_format: big-nc
```

Though, as discussed in Jira, we will define a version per format.